client.go

```diff
@@ -1374,8 +1374,8 @@ func (cl *Client) allTorrentsCompleted() bool {
 // Returns true when all torrents are completely downloaded and false if the
 // client is stopped before that.
 func (cl *Client) WaitAll() bool {
-	cl.lock()
-	defer cl.unlock()
+	cl.rLock()
```

Comment:
I don't think you can RLock if you plan to use Cond.Wait, as it wants to Unlock. The other changes look good though.

Reply:
yes. fixed